src/main/java/com/yahoo/bullet/storm/QueryBolt.java

```diff
@@ -5,139 +5,148 @@
-        if (!metadataKeys.isEmpty()) {
-            log.info("Metadata collection is enabled");
-            log.info("Collecting metadata for these concepts:\n{}", metadataKeys);
-            // Add all metadataKeys back to configuration for reuse so no need refetch them on every new query
-            configuration.put(BulletStormConfig.RESULT_METADATA_METRICS_MAPPING, metadataKeys);
-        }
-
+        queries = new HashMap<>();
+        classifier = new TupleClassifier();
         // Enable built in metrics
-        metricsEnabled = (Boolean) configuration.getOrDefault(BulletStormConfig.TOPOLOGY_METRICS_BUILT_IN_ENABLE,
-                                                              DEFAULT_BUILT_IN_METRICS_ENABLE);
-        metricsIntervalMapping = (Map<String, Number>) configuration.getOrDefault(BulletStormConfig.TOPOLOGY_METRICS_BUILT_IN_EMIT_INTERVAL_MAPPING,
-                                                                                  new HashMap<>());
-        metricsIntervalMapping.putIfAbsent(DEFAULT_BUILT_IN_METRICS_INTERVAL_KEY, DEFAULT_BUILT_IN_METRICS_INTERVAL_SECS);
+        metricsEnabled = config.getAs(BulletStormConfig.TOPOLOGY_METRICS_BUILT_IN_ENABLE, Boolean.class);
+        metricsIntervalMapping = config.getAs(BulletStormConfig.TOPOLOGY_METRICS_BUILT_IN_EMIT_INTERVAL_MAPPING, Map.class);
+    }
+
+    @Override
+    public void cleanup() {
     }
 
     /**
-     * Retires queries that are active past the tick time.
+     * Handles a metadata message for a query.
      *
-     * @return The map of query ids to queries that were retired.
+     * @param tuple The metadata tuple.
+     * @return The created {@link Metadata}.
      */
-    protected Map<String, Q> retireQueries() {
-        Map<String, Q> retiredQueries = queriesMap.entrySet().stream().filter(e -> e.getValue().isExpired())
-                                                                      .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
-        queriesMap.keySet().removeAll(retiredQueries.keySet());
-        if (retiredQueries.size() > 0) {
-            log.info("Retired queries: {}. Active queries: {}.", retiredQueries.size(), queriesMap.size());
+    protected Metadata onMeta(Tuple tuple) {
+        String id = tuple.getString(TopologyConstants.ID_POSITION);
+        Metadata metadata = (Metadata) tuple.getValue(TopologyConstants.METADATA_POSITION);
+        if (metadata == null) {
+            return null;
         }
-        return retiredQueries;
+        Metadata.Signal signal = metadata.getSignal();
+        if (signal == Metadata.Signal.KILL || signal == Metadata.Signal.COMPLETE) {
+            removeQuery(id);
+            log.info("Received {} signal and killed query: {}", signal, id);
+        }
+        return metadata;
     }
 
     /**
-     * Initializes a query from a query tuple.
+     * For testing only. Create a {@link Querier} from the given query ID, body and configuration.
```

Comment:
I saw this method is been called in FilterBolt and JoinBolt. not testing code. Is it the same one?

Reply:
Yes this is used because this is the easiest way to inject a custom (mocked or spied) Querier when testing. I extend FilterBolt or JoinBolt and override this method to do that in testing.